### DISCUSSION TITLE
Migrating to v2

### DISCUSSION THREAD

**Discussion by BrandonALittle** [parcel-bundler/parcel#5416] [General] [answered]
The documentation for migration from v1 to v2 does not explain how to migrate. There are major differences in config between the two, especially when using the API in a script, but I can't find any explanation about how the configuration for v2 is different from v1.

Could you please update your docs with information that would help someone to migrate?

**Comment by mischnic** [comment, marked as answer]
You mean https://v2.parceljs.org/getting-started/migration/?

> configuration for v2 is different from v1

What exactly do you mean by configuration?

**Reply by BrandonALittle** [reply]
I use a script in which I am attempting to construct a Parcel, following the example: 
```
(async () => {
  let bundler = new Parcel({
    entries: path.join(__dirname, "src/index.js"),
    defaultConfig: require.resolve("@parcel/config-default"),
    defaultEngines: {
      browsers: ["last 1 Chrome version"],
      node: "10",
    },
    mode: "production",
  });

  await bundler.run();
})();
```

This seems very different from how the Bundler class from v1 works. Options are different, perhaps renamed. I was previously passing an array of entry files and an options object to the Bundler constructor. 

By configuration, I mean the entire options object to pass to the Parcel constructor as well as the config and defaultConfig properties on that object.

That migration link you shared makes it seem like v2 will work the same as v1, with those minor gotchas mentioned on that page. It doesn't, however, mention anything regarding how Parcel is different from Bundler

The only clue I've found is the InitialParcelOptions type, but the types there don't clearly match up with the options I'm using in my v1 script.

**Reply by mischnic** [reply]
~Have you seen this: https://v2.parceljs.org/features/parcel-api/ ?~ You have

But you're right, that's missing in the migration guide (-> https://github.com/parcel-bundler/website/issues/778)

**Reply by mischnic** [reply]
> The only clue I've found is the InitialParcelOptions type, but the types there don't clearly match up with the options I'm using in my v1 script.

Happy to answer questions in the meantime. I think these options mostly stayed the same? What are you having trouble with?

**Reply by BrandonALittle** [reply]
One problem I'm having is that I have to use `const Parcel = require('@parcel/core').default` to access the constructor. Not sure why the import syntax won't work for me, but I think that's an issue on my end. 

Trying to start simple, I've just modified the entries property, but this: 
```
(async () => {
  let bundler = new Parcel({
    entries: ['./src/Index.tsx'],
    defaultConfig: require.resolve("@parcel/config-default"),
    defaultEngines: {
      browsers: ["last 1 Chrome version"],
      node: "10",
    },
    mode: "production",
  });

  await bundler.run();
})();
```

throws an error: `TypeError [ERR_INVALID_ARG_TYPE]: The "to" argument must be of type string. Received type undefined`

**Reply by mischnic** [reply]
I've tried that on my machine and it works (new folder, `yarn add parcel@nightly`, adding a `index.tsx` and a `run.js` for your snippet). Can you post the stack trace of the error?

**Reply by BrandonALittle** [reply]
Hmmm. Then I'm wondering if I have an issue with my Babel or Typescript setup...
Here's the stack trace:
```
(node:79074) UnhandledPromiseRejectionWarning: TypeError [ERR_INVALID_ARG_TYPE]: The "to" argument must be of type string. Received type undefined
    at validateString (internal/validators.js:125:11)
    at Object.relative (path.js:1163:5)
    at processConfigChain (../node_modules/@parcel/core/lib/loadParcelConfig.js:153:36)
    at create (../node_modules/@parcel/core/lib/loadParcelConfig.js:66:10)
    at loadParcelConfig (../node_modules/@parcel/core/lib/loadParcelConfig.js:42:26)
```

**Reply by BrandonALittle** [reply]
Ya, so it looks like I needed to add the `.parcelrc` file, which I wasn't using previously. After copying the snippet from your docs: 
```
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.{ts,tsx}": ["@parcel/transformer-typescript-tsc"]
  }
}
```
..the bundler gets past that error, but now the compiler doesn't like `SyntaxError: Unexpected token export` for module exports.... Probably another TS issue.

**Reply by mischnic** [reply]
> compiler doesn't like SyntaxError: Unexpected token export for module exports.... Probably another TS issue.

Please try installing all parcel packages in the `nightly` version. I think there was a publishing issue with the alpha of `@parcel/transformer-typescript-tsc` where we didn't transpile the sources.

**Reply by BrandonALittle** [reply]
After `yarn add -D parcel@nightly`, I'm still getting the same error.
```
SyntaxError: Unexpected token export
    at Module._compile (internal/modules/cjs/loader.js:721:23)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at NodePackageManager.load (../node_modules/@parcel/package-manager/lib/NodePackageManager.js:192:9)
    at NodePackageManager.requireSync (../node_modules/@parcel/package-manager/lib/NodePackageManager.js:156:17)
    at Module.m.require.id [as require] (../node_modules/@parcel/package-manager/lib/NodePackageManager.js:178:19)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (../node_modules/@parcel/transformer-typescript-tsc/lib/TSCTransformer.js:10:16)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
```

**Reply by mischnic** [reply]
`@parcel/transformer-typescript-tsc` wouldn't be installed when installing `parcel`, so this is probably listed in your project's dependencies as well (with an explicit version).

**Reply by BrandonALittle** [reply]
Thanks! That definitely resolved that error. Getting some others, but I'll dig in and let you know what I find.

**Reply by BrandonALittle** [reply]
I'm seeing a lot of issues with tilde `~` module resolution, but I noticed a comment in the docs referencing an issue with that 'paths' property in `tsconfig`. Is the tilde altogether prohibited with v2 when using the `@parcel/transformer-typescript-tsc`?

**Reply by BrandonALittle** [reply]
I managed to get the bundler to run without error, but it doesn't seem to be bundling my node modules. In v1 there was a config option to enable this, but in v2 I don't see a way to do the same. I tried activating the bundle analyzer as described here: https://v2.parceljs.org/features/production/ but that's not working as expected either. I'm still working on it...

**Reply by mischnic** [reply]
The tilde might point to a different folder now: https://v2.parceljs.org/features/module-resolution/#tilde-paths

> but it doesn't seem to be bundling my node modules

Are you intentionally bundling for Node? If not, do you have `engines: {node: ...}` in your package.json or `main: "index.js"`?
Otherwise: there's `includeNodeModules` in the target options: https://v2.parceljs.org/configuration/package-json/#includenodemodules

> I tried activating the bundle analyzer as described here: v2.parceljs.org/features/production but that's not working as expected either

What have you tried but didn't work?

**Reply by BrandonALittle** [reply]
Yes, I'm intentionally building for Node. I did have the `engines: {node: ...}`, and tried removing that without success. I do not have a `targets` entry anywhere, so I'm just wondering why that is necessary with v2 but not v1. 

As far as the bundle analyzer, I added the environment variable as instructed at the Production page, but there's no output. I suspect it's just failing before getting to that point.

**Reply by mischnic** [reply]
> Yes, I'm intentionally building for Node

Then having the engines is fine.

> I do not have a targets entry anywhere, so I'm just wondering why that is necessary with v2 but not v1.

You can also pass targets via the API: https://v2.parceljs.org/plugin-system/api/#InitialParcelOptions. 
With V1, `bundleNodeModules` was a global option, now it was renamed and moved to the target specific options. Something like
```js
targets: {
  "default": {
     "includeNodeModules": true
  }
}
```
should work.


> As far as the bundle analyzer, I added the environment variable as instructed at the Production page, but there's no output. I suspect it's just failing before getting to that point.

That is printed on the `buildSuccess`, so if you get a build error before that, there will be no report.

**Comment by BrandonALittle** [comment]
Sorry for the delay. Using the example snippet with the addition of the distDir, I managed to get the bundle to build. But two issues remain:
1. I expected the v2 bundle to be smaller than the v1 bundle, but that is not the case. Perhaps tree-shaking is not enabled?
2. I am not getting any output from the bundle-analyzer, and no parce-bundle-reports folder is being generated.

**Reply by BrandonALittle** [reply]
I am currently passing `scopeHoist: false` in the options, so it looks like that disables tree-shaking. Changing this option to `true` results in a JS Heap out of memory error.

**Reply by BrandonALittle** [reply]
Resolved by changing version of Node from v10 to v14
